Marlin/src/module/stepper.cpp

```diff
@@ -2091,45 +2113,11 @@ hal_timer_t Stepper::calc_timer_interval(uint32_t step_rate) {
 
 // Get the timer interval and the number of loops to perform per tick
 hal_timer_t Stepper::calc_multistep_timer_interval(uint32_t step_rate) {
-  #if MULTISTEPPING_LIMIT == 1
-
-    // Just make sure the step rate is doable
-    NOMORE(step_rate, uint32_t(MAX_STEP_ISR_FREQUENCY_1X));
-
-  #else
-
-    // The stepping frequency limits for each multistepping rate
-    static const uint32_t limit[] PROGMEM = {
-          (  MAX_STEP_ISR_FREQUENCY_1X     )
-        , (  MAX_STEP_ISR_FREQUENCY_2X >> 1)
-      #if MULTISTEPPING_LIMIT >= 4
-        , (  MAX_STEP_ISR_FREQUENCY_4X >> 2)
-      #endif
-      #if MULTISTEPPING_LIMIT >= 8
-        , (  MAX_STEP_ISR_FREQUENCY_8X >> 3)
-      #endif
-      #if MULTISTEPPING_LIMIT >= 16
-        , ( MAX_STEP_ISR_FREQUENCY_16X >> 4)
-      #endif
-      #if MULTISTEPPING_LIMIT >= 32
-        , ( MAX_STEP_ISR_FREQUENCY_32X >> 5)
-      #endif
-      #if MULTISTEPPING_LIMIT >= 64
-        , ( MAX_STEP_ISR_FREQUENCY_64X >> 6)
-      #endif
-      #if MULTISTEPPING_LIMIT >= 128
-        , (MAX_STEP_ISR_FREQUENCY_128X >> 7)
-      #endif
-    };
-
-    // Find a doable step rate using multistepping
-    uint8_t multistep = 1;
-    for (uint8_t i = 0; i < COUNT(limit) && step_rate > uint32_t(pgm_read_dword(&limit[i])); ++i) {
-      step_rate >>= 1;
-      multistep <<= 1;
-    }
-    steps_per_isr = multistep;
-
+  #if MULTISTEPPING_LIMIT > 1
+    uint8_t loops = steps_per_isr;
+    if (MULTISTEPPING_LIMIT >= 16 && loops >= 16) { step_rate >>= 4; loops >>= 4; }
+    if (MULTISTEPPING_LIMIT >=  4 && loops >=  4) { step_rate >>= 2; loops >>= 2; }
+    if (MULTISTEPPING_LIMIT >=  2 && loops >=  2) { step_rate >>= 1; }
```

Comment:
Quick question: Is this approach as accurate as the `limit` table? I have a vague suspicion that the `(F_CPU) / ISR_EXECUTION_CYCLES(n)` values might have one or more extra digits that carry, and those might be getting lost if we are just shifting them out. But, I assume you already checked.

Reply:
From the perspective of `step_rate`, which this code is calculating, there is no change. It was right shifted by a certain number of bits in the old code and that's still the case in the new code. What is different is how we derive the number of bits to shift by.